mne/time_frequency/csd.py

```diff
@@ -1,56 +1,512 @@
+            requested frequency or frequency-bin.
+
+        See Also
+        --------
+        get_data
+        """
+        if freq is None and index is None:
+            raise ValueError('Use either the "freq" or "index" parameter to '
+                             'select the desired frequency.')
+
+        elif freq is not None:
+            if index is not None:
+                raise ValueError('Cannot specify both a frequency and index.')
+
+            index = self._get_frequency_index(freq)
+
+        return self[index]
+
+    def get_data(self, frequency=None, index=None):
+        """Get the CSD matrix for a given frequency as NumPy array.
+
+        If there is only one matrix defined in the CSD object, calling this
+        method without any parameters will return it. If multiple matrices are
+        defined, use either the ``frequency`` or ``index`` parameter to select
+        one.
+
+        Parameters
+        ----------
+        frequency : float | None
+            Return the CSD matrix for a specific frequency. Only available when
+            no averaging across frequencies has been done.
+        index : int | None
+            Return the CSD matrix for the frequency or frequency-bin with the
+            given index.
+
+        Returns
+        -------
+        csd : ndarray, shape (n_channels, n_channels)
+            The CSD matrix corresponding to the requested frequency.
+
+        See Also
+        --------
+        pick_frequency
+        """
+        if frequency is None and index is None:
+            if self._data.shape[1] > 1:
+                raise ValueError('Specify either the frequency or index of '
+                                 'the frequency bin for which to obtain the '
+                                 'CSD matrix.')
+            index = 0
+        elif frequency is not None:
+            if index is not None:
+                raise ValueError('Cannot specify both a frequency and index.')
+            index = self._get_frequency_index(frequency)
+
+        return _vector_to_sym_mat(self._data[:, index])
+
+    @copy_function_doc_to_method_doc(plot_csd)
+    def plot(self, info=None, mode='csd', colorbar=True, cmap='viridis',
+             n_cols=None, show=True):
+        return plot_csd(self, info=info, mode=mode, colorbar=colorbar,
+                        cmap=cmap, n_cols=n_cols, show=show)
+
+    def __setstate__(self, state):  # noqa: D105
+        self._data = state['data']
+        self.tmin = state['tmin']
+        self.tmax = state['tmax']
+        self.ch_names = state['ch_names']
+        self.frequencies = state['frequencies']
+        self.n_fft = state['n_fft']
+
+    def __getstate__(self):  # noqa: D105
+        return dict(
+            data=self._data,
+            tmin=self.tmin,
+            tmax=self.tmax,
+            ch_names=self.ch_names,
+            frequencies=self.frequencies,
+            n_fft=self.n_fft,
+        )
```

Comment:
why are these two methods necessary?

Reply:
To save and read from/to an HDF5 file, we need a dictionary representation of the object. This could just be done in the `save` method and `read_csd` function, but if we do this in a separate `__setstate__` and `__getstate__` function, than `pickle` can also make use of this (and maybe deepcopy too?).

Reply:
ok ok